server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpgradeService.java

```diff
@@ -93,6 +104,87 @@ public IndexMetadata upgradeIndexMetadata(IndexMetadata indexMetadata,
+
+                newMetadata.setTable(
+                    docTable.versionCreated().before(DocTableInfo.COLUMN_OID_VERSION)
+                        ? NO_OID_COLUMN_OID_SUPPLIER
+                        : newMetadata.columnOidSupplier(),
+                    relationName,
+                    docTable.allReferences(),
+                    // If RelationMetadata exist in the cluster state, make sure to override them with
+                    // the upgraded settings which currently takes place on IndexMetadata
+                    indexMetadata.getSettings(),
+                    docTable.clusteredBy(),
+                    docTable.columnPolicy(),
+                    docTable.pkConstraintName(),
+                    docTable.checkConstraints()
+                        .stream().collect(Collectors.toMap(CheckConstraint::name, CheckConstraint::expressionStr)),
+                    docTable.primaryKey(),
+                    docTable.partitionedBy(),
+                    indexMetadata.getState(),
+                    indexUUIDs
+                );
+            }
+        }
+
+        for (ObjectObjectCursor<String, IndexTemplateMetadata> entry : metadata.templates()) {
+            IndexTemplateMetadata indexTemplateMetadata = entry.value;
+            RelationName relationName = IndexName.decode(indexTemplateMetadata.name()).toRelationName();
+            DocTableInfo docTable = tableFactory.create(relationName, metadata);
+            List<String> indexUUIDs = metadata.getIndices(
+                relationName,
+                List.of(),
+                false,
+                IndexMetadata::getIndexUUID
+            );
+            // versionCreated could be missing from the template settings and "calculated" afterwards
+            // in DocTableInfo, so put it back to RelationMetadata#parameters
+            Settings settings = Settings.builder()
+                // If RelationMetadata exist in the cluster state, make sure to override them with
+                // the upgraded settings which currently takes place on IndexTemplateMetadata
+                .put(indexTemplateMetadata.settings())
+                .put(SETTING_INDEX_VERSION_CREATED.getKey(), docTable.versionCreated())
+                .build();
+            newMetadata.setTable(
+                docTable.versionCreated().before(DocTableInfo.COLUMN_OID_VERSION)
+                    ? NO_OID_COLUMN_OID_SUPPLIER
+                    : newMetadata.columnOidSupplier(),
+                relationName,
+                docTable.allReferences(),
+                settings,
+                docTable.clusteredBy(),
+                docTable.columnPolicy(),
+                docTable.pkConstraintName(),
+                docTable.checkConstraints()
+                    .stream().collect(Collectors.toMap(CheckConstraint::name, CheckConstraint::expressionStr)),
+                docTable.primaryKey(),
+                docTable.partitionedBy(),
+                IndexMetadata.State.OPEN,
```

Comment:
Is it possible that the template has state=closed?

https://github.com/crate/crate/blob/99e91b665f8b7c3dfe4008116d246dc64b987a60/server/src/main/java/io/crate/metadata/doc/DocTableInfoFactory.java#L258-L260